VERSIONING.md

```diff
@@ -0,0 +1,71 @@
+
+### Examples of MAJOR Command-Line Changes
+
+- Complete removal of an option.
+- Changing the behavior of an option so that it no longer does what it was defined to do.
+- Removing the ability to run under a particular platform.
+
+### Examples of MINOR Command-Line Changes
+
+- Addition of a new option
+- Modifying the behavior of an option so it does the same thing better or more effectively or has added sub-options.
+- Adding support for running under an new platform
+
+### Examples of PATCH Command-Line changes
+
+- Fixing a bug in the behavior of an option.
+- Restoring an option to its original defined behavior.
+- Correcting spelling errors.
+
+## Engine APIs
+
+The **NUnit Engine** is intended to be used by third-party runners as well as our own **Console Runner**. Usage is through the defined API, specified by interfaces included in the `nunit.engine.api` assembly. It's important to note that public types and methods included in other assemblies are not considered a part of the published API. Direct creation and use of `nunit.engine`, `nunit.engine.core` or any other assemblies included in our packages is discouraged for that reason. See below for more about this.
+
+### Examples of MAJOR Api Changes
+
+- Removing a published interface.
+- Modifying an interface so that it breaks existing usage.
+
+### Examples of MINOR Api Changes
+
+- Adding a new published interface and implementation.
+- Modifying an interface in a way which does __not__ break existing usage.
+- Modifying the implementation of an interface so it does the same thing better or more effectively.
+- Restoring an API or it's implementation to the original behavior following an erroneous MINOR release.
+
+### Examples of PATCH Api Changes
+
+- Fixing bugs in the implementation of an API.
+- Restoring an API or it's implementation to the original behavior following an erroneous PATCH release.
+
+## Notes
+
+1. Access to Types and Methods through reflection, while possible, is not part of our supported API.
+
+2. As explained above, the existing `nunit.engine.api` assembly specifies the _entire_ published API for the **NUnit Engine**. Other guaranteed APIs may be added in the future, using the same `.api` naming convention.
+
+3. Historically, NUnit has used public visibility for many Types and Methods, which were not actually part of the API. We are in the process of changing this approach to a more modern focus on limiting visibility. That said, even where a Type or Method is publicly accessible, it is __not__ automatically a part of our API and may change from release to release.
```

Comment:
Even in the `.api` assembly? Since that's the only assembly where the visibility of types could matter?

Reply:
I'll look at the wording.

In the case of the API assembly, I think we would be hard pressed to argue that making something public in an assembly called "API" isn't "publishing" it. So we need to eliminate any public stuff we don't want to support. A great example of this is the public constructor of TestPackage that takes a string. It was never intended to be used except in a specific situation internally and causes a lot of work for us to maintain tests using both forms of constructor.

However, I did want to restate this for any users who use public methods out of other assemblies. Belt and suspenders, I guess. :-)